### PR TITLE
fix(web): roll back file editor when save fails

### PR DESCRIPTION
**Pull request by Adolanium**
## What Changed

The file editor now restores the last confirmed text when a save fails.

Edits still apply to the overlay first. If the latest write fails, the overlay goes back, the unsaved mark clears, and a toast says the save failed. A newer edit that landed during that write is not rolled back. The coordinator retries it.

## Why

A failed write left the new text on screen. Reload then showed the old disk contents. Easy to think the file was saved.

## UI Changes

Failed saves now toast and restore the last confirmed file text. No other visual change.

## Checklist

- [x] This PR is small and focused
- [x] I explained what changed and why
- [x] I included before/after screenshots for any UI changes (N/A. toast plus restored text is the change)
- [x] I included a video for animation/interaction changes (N/A)

---

> [!NOTE]
> **Medium Risk**
> Changes autosave and optimistic file cache behavior on write failure; incorrect rollback could drop user edits or show stale disk content, though guards exist for newer overlay edits and interrupted commands.
> 
> **Overview**
> Failed file writes no longer leave optimistic editor text on screen. **`FileSaveCoordinator`** now tracks **`lastConfirmedContents`** (from **`initialContents`** and successful saves), exposes **`syncConfirmed`** so idle server refreshes update the rollback baseline, and on a non-interrupted persist failure resets internal state and invokes **`onRollback`** instead of leaving the file “saved.” Interrupted writes and in-flight newer edits are unchanged: stale failures do not roll back newer typing, and overlapping saves still debounce to the latest revision.
> 
> **`FilePreviewPanel`** wires this up by passing loaded **`contents`**, syncing confirmed text when **`contents`** changes without rebuilding the coordinator on every keystroke, and implementing **`onRollback`** to restore project file query data, clear pending when appropriate, skip rollback if the overlay already diverged, and show a **“Could not save file”** error toast.
> 
> Tests cover rollback, baseline refresh, interrupted/dispose edge cases, and concurrent edit behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit de49ca51f059cc5e0c5701fc5431208987af8ba5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Roll back file editor to last confirmed contents on save failure
> - `FileSaveCoordinator` now tracks a `lastConfirmedContents` baseline, seeded from a new `initialContents` option and updated on successful saves and idle external refreshes via `syncConfirmed`.
> - On save failure, the coordinator restores `latestContents` to the baseline, resets `latestRevision` to 0, and invokes a new `onRollback` callback with `failedContents`, `confirmedContents`, and the persist `result`; pending state is intentionally left uncleared so the UI can mark the file as pending.
> - Write interruptions detected by `isAtomCommandInterrupted` are treated as a distinct case — no rollback, no pending change.
> - `useFileSaveCoordinator` wires `initialContents` and a new `useEffect` that calls `coordinator.syncConfirmed(contents)` on external content changes, and implements `onRollback` to reset query data and show an error toast.
> - Behavioral Change: pending state is no longer cleared on save failure or interruption; callers that relied on pending being cleared after a failed save will need updating.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized de49ca5.</sup>
> 
>

### CODE REVIEW

File: apps/web/src/components/files/FilePreviewPanel.tsx [hunk 1]
```diff
@@ -423,7 +425,17 @@ function useFileSaveCoordinator({
...
+          toastManager.add({
+            type: "error",
+            title: "Could not save file",
+            description: relativePath,
+          });
+        },
       }),
+    // initialContents is the loaded file at mount. Overlay edits must not
+    // rebuild the coordinator or last-confirmed state resets on every keystroke.
     [cwd, environmentId, onPendingChange, relativePath, writeFile],
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx:447]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/FilePreviewPanel.tsx [hunk 2]
```diff
@@ -423,7 +425,17 @@ function useFileSaveCoordinator({
         onConfirmed: (confirmedContents) => {
           confirmProjectFileQueryData(environmentId, cwd, relativePath, confirmedContents);
         },
+        onRollback: (confirmedContents) => {
+          setProjectFileQueryData(environmentId, cwd, relativePath, confirmedContents);
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx:434]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx:434]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/fileSaveCoordinator.ts [hunk 3]
```diff
@@ -51,15 +56,24 @@ export class FileSaveCoordinator<A = unknown, E = unknown> {
...
+      succeeded = false;
     }
 
     this.saving = false;
     if (revision === this.latestRevision) {
-      if (succeeded) this.options.onPendingChange(false);
+      if (!succeeded) {
+        this.options.onRollback(this.lastConfirmedContents);
+      }
+      this.options.onPendingChange(false);
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/fileSaveCoordinator.ts]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/fileSaveCoordinator.ts [hunk 4]
```diff
@@ -51,15 +56,24 @@ export class FileSaveCoordinator<A = unknown, E = unknown> {
...
+    } catch {
+      succeeded = false;
     }
 
     this.saving = false;
     if (revision === this.latestRevision) {
-      if (succeeded) this.options.onPendingChange(false);
+      if (!succeeded) {
+        this.options.onRollback(this.lastConfirmedContents);
+      }
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/fileSaveCoordinator.ts]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/FilePreviewPanel.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/FilePreviewPanel.tsx [hunk 5]
```diff
@@ -423,7 +425,17 @@ function useFileSaveCoordinator({
...
+          toastManager.add({
+            type: "error",
+            title: "Could not save file",
+            description: relativePath,
+          });
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx:442]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/fileSaveCoordinator.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/fileSaveCoordinator.ts]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/fileSaveCoordinator.ts]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/FilePreviewPanel.tsx [hunk 6]
```diff
@@ -423,7 +425,17 @@ function useFileSaveCoordinator({
...
+        },
       }),
+    // initialContents is the loaded file at mount. Overlay edits must not
+    // rebuild the coordinator or last-confirmed state resets on every keystroke.
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx:446]: Fixed on a58ea12.

Interrupted writes no longer roll back or toast. A disposed coordinator no longer rolls back or clears pending. A failed latest write resets the revision so dispose cannot flush the discarded text.

Rollback now clears the overlay only if it still matches the failed write. Idle query refreshes update the confirmed baseline. The toast uses the write error.

File: apps/web/src/components/files/fileSaveCoordinator.ts [hunk 7]
```diff
@@ -51,15 +71,41 @@ export class FileSaveCoordinator<A = unknown, E = unknown> {
...
+        this.latestRevision = 0;
+        this.latestContents = this.lastConfirmedContents;
+        if (!this.disposed) {
+          this.options.onRollback({
+            failedContents: contents,
+            confirmedContents: this.lastConfirmedContents,
+            result,
+          });
+          this.options.onPendingChange(false);
+        }
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/fileSaveCoordinator.ts]: Fixed on dc19c82. A failed write that finishes after dispose still rolls back and clears pending. If a newer overlay is already there, that text stays and the file is marked pending again.

File: apps/web/src/components/files/FilePreviewPanel.tsx [hunk 8]
```diff
@@ -423,10 +426,28 @@ function useFileSaveCoordinator({
         onConfirmed: (confirmedContents) => {
           confirmProjectFileQueryData(environmentId, cwd, relativePath, confirmedContents);
         },
+        onRollback: ({ failedContents, result }) => {
+          const overlay = getOptimisticProjectFileQueryData(environmentId, cwd, relativePath);
+          if (overlay !== null && overlay.contents !== failedContents) {
+            return;
+          }
+          clearProjectFileQueryData(environmentId, cwd, relativePath);
+          const error = result === null ? null : squashAtomCommandFailure(result);
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx]: Fixed on 236f529. Rollback owns pending now, so a newer overlay keeps the unsaved mark. Failed saves restore last confirmed text, then confirm it. The toast only squashes Failure results.

File: apps/web/src/components/files/FilePreviewPanel.tsx [hunk 9]
```diff
@@ -423,10 +426,28 @@ function useFileSaveCoordinator({
         onConfirmed: (confirmedContents) => {
           confirmProjectFileQueryData(environmentId, cwd, relativePath, confirmedContents);
         },
+        onRollback: ({ failedContents, result }) => {
+          const overlay = getOptimisticProjectFileQueryData(environmentId, cwd, relativePath);
+          if (overlay !== null && overlay.contents !== failedContents) {
+            return;
+          }
+          clearProjectFileQueryData(environmentId, cwd, relativePath);
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx]: Fixed on 236f529. Rollback owns pending now, so a newer overlay keeps the unsaved mark. Failed saves restore last confirmed text, then confirm it. The toast only squashes Failure results.

File: apps/web/src/components/files/fileSaveCoordinator.ts [hunk 10]
```diff
@@ -51,15 +71,39 @@ export class FileSaveCoordinator<A = unknown, E = unknown> {
...
+      }
+      if (!succeeded) {
+        this.latestRevision = 0;
+        this.latestContents = this.lastConfirmedContents;
+        this.options.onRollback({
+          failedContents: contents,
+          confirmedContents: this.lastConfirmedContents,
+          result,
+        });
+        this.options.onPendingChange(false);
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/fileSaveCoordinator.ts]: Fixed on 236f529. Rollback owns pending now, so a newer overlay keeps the unsaved mark. Failed saves restore last confirmed text, then confirm it. The toast only squashes Failure results.

File: apps/web/src/components/files/FilePreviewPanel.tsx [hunk 11]
```diff
@@ -423,10 +426,29 @@ function useFileSaveCoordinator({
         onConfirmed: (confirmedContents) => {
           confirmProjectFileQueryData(environmentId, cwd, relativePath, confirmedContents);
         },
+        onRollback: ({ failedContents, result }) => {
+          const overlay = getOptimisticProjectFileQueryData(environmentId, cwd, relativePath);
+          if (overlay !== null && overlay.contents !== failedContents) {
+            onPendingChange(relativePath, true);
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/FilePreviewPanel.tsx:431]: Fixed on 236f529. Rollback owns pending now, so a newer overlay keeps the unsaved mark. Failed saves restore last confirmed text, then confirm it. The toast only squashes Failure results.

File: apps/web/src/components/files/fileSaveCoordinator.ts [hunk 12]
```diff
@@ -51,15 +71,38 @@ export class FileSaveCoordinator<A = unknown, E = unknown> {
...
+        this.latestRevision = 0;
+        this.latestContents = this.lastConfirmedContents;
+        this.options.onRollback({
+          failedContents: contents,
+          confirmedContents: this.lastConfirmedContents,
+          result,
+        });
+        return;
+      }
+      if (!this.disposed) this.options.onPendingChange(false);
```
[the remark this answers is not in the document]
Comment (Adolanium) [reply apps/web/src/components/files/fileSaveCoordinator.ts:106]: Fixed on de49ca5. After a successful latest save, latestRevision goes back to 0 so idle refreshes can update the confirmed baseline.
